File: LRSMessageToolBar/Classes/InputBar/LRSMessageInputBar.m
//
//  LRSMessageToolsBar.m
//  LRSMessageToolBar
//
//  Created by sama 刘 on 2021/10/27.
//

#import "LRSMessageInputBar.h"
#import <MASonry/MASonry.h>
#import "LRSMessageToolBarHelper.h"
#import "LRSMessageToolBarConfigure.h"

@interface LRSMessageInputBar()
@property (nonatomic, strong) LRSMessageToolBarConfigure *configure;
@property (nonatomic, strong) MASConstraint *textViewHeight;
@end

@implementation LRSMessageInputBar
@synthesize mode = _mode;

+ (instancetype)toolBarWithConfigure:(LRSMessageToolBarConfigure *)configure {
    LRSMessageInputBar *bar = [[self alloc] init];
    bar.configure = configure;
    [bar loadBasicSubviews];
    [bar configureSubviews];
    bar.mode = LRSMessageToolBarModeTextInput;
    return bar;
}

- (void)configureSubviews {
    [self.recordingBtn setBackgroundImage:[self.configure.recordConfigure.stateConfigure imageForState:UIControlStateNormal] forState:UIControlStateNormal];
    [self.recordingBtn setBackgroundImage:[self.configure.recordConfigure.stateConfigure imageForState:UIControlStateSelected] forState:UIControlStateSelected];
    [self.recordingBtn setAttributedTitle:[self.configure.recordConfigure.stateConfigure titleForState:UIControlStateNormal] forState:UIControlStateNormal];
    [self.recordingBtn setAttributedTitle:[self.configure.recordConfigure.stateConfigure titleForState:UIControlStateSelected] forState:UIControlStateSelected];

    [self.imagePickButton setImage:[self.configure.buttonConfigure.imagePickerButtonStateConfigure imageForState:UIControlStateNormal] forState:(UIControlStateNormal)];

    [self.faceButton setImage:[self.configure.buttonConfigure.emojiButtonStateConfigure imageForState:UIControlStateNormal] forState:(UIControlStateNormal)];

    self.inputTextView.placeHolder = self.configure.textViewConfigure.placeholder;

    self.recordingBtn.areaY = self.configure.textViewConfigure.minHeight;

}

- (void)loadBasicSubviews {
    // 底部视图容器
    [self addSubview:self.imagePickButton];
    [self addSubview:self.faceButton];
    [self addSubview:self.inputTextView];
    [self addSubview:self.recordingBtn];
    [self addSubview:self.modeSwitchButton];
    [self addSubview:self.bottomLine];

    [self.bottomLine mas_makeConstraints:^(MASConstraintMaker *make) {
        make.bottom.equalTo(self.inputTextView.mas_bottom).offset(-1);
        make.left.right.equalTo(self.inputTextView);
        make.height.equalTo(@0.5);
    }];
    [self.modeSwitchButton mas_remakeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(@(self.configure.modeSwitchConfigure.leftMargin));
        make.size.mas_equalTo(self.configure.modeSwitchConfigure.buttonSize);
        make.centerY.equalTo(self.inputTextView);
    }];
    [self.recordingBtn mas_remakeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(self.modeSwitchButton.mas_right).offset(self.configure.textViewConfigure.leftMargin);
        make.top.mas_equalTo(self.configure.textViewConfigure.topMargin);
        make.right.equalTo(self.imagePickButton.mas_left).offset(-self.configure.textViewConfigure.topMargin);
        make.height.mas_equalTo(self.configure.textViewConfigure.minHeight);
    }];
    [self.inputTextView mas_remakeConstraints:^(MASConstraintMaker *make) {
        make.left.top.equalTo(self.recordingBtn);
        make.right.equalTo(self.faceButton.mas_left).offset(-self.configure.textViewConfigure.topMargin);
        self.textViewHeight = make.height.mas_equalTo(self.configure.textViewConfigure.minHeight);
    }];
    [self.imagePickButton mas_remakeConstraints:^(MASConstraintMaker *make) {
        make.right.equalTo(self.mas_right).offset(-self.configure.buttonConfigure.rightButtonsMargin);
        make.centerY.equalTo(self.modeSwitchButton);
        make.width.height.equalTo(self.modeSwitchButton);
    }];
    [self.faceButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.size.equalTo(self.imagePickButton);
        make.centerY.equalTo(self.imagePickButton);
        make.right.equalTo(self.imagePickButton.mas_left).offset(self.configure.buttonConfigure.rightButtonsXSpacing);
    }];
}

- (void)setMode:(LRSMessageToolBarMode)mode {
    _mode = mode;
    [self onModeSwitch:mode];
}

- (void)modeSwitchClick:(UIButton *)button {
    self.mode = !_mode;
}

- (void)onModeSwitch:(LRSMessageToolBarMode)mode {
    self.recordingBtn.hidden = mode == LRSMessageToolBarModeTextInput;
    self.modeSwitchButton.selected = !self.recordingBtn.hidden;
    self.inputTextView.hidden = !self.recordingBtn.hidden;
    self.faceButton.hidden = self.inputTextView.hidden;
    [self.modeSwitchButton setBackgroundImage:[self.configure.modeSwitchConfigure.stateConfigure imageForState:mode == LRSMessageToolBarModeTextInput ? UIControlStateNormal : UIControlStateSelected] forState:UIControlStateNormal];

}

- (UIView *)bottomLine {
    if (_bottomLine == nil) {
        _bottomLine = [[UIView alloc] init];
        _bottomLine.backgroundColor = [UIColor colorWithRed:226/255.0 green:226/255.0 blue:226/255.0 alpha:1];
    }
    return _bottomLine;
}

- (LRSPlaceholderTextView *)inputTextView {
    if (!_inputTextView) {
        _inputTextView = [[LRSPlaceholderTextView alloc] initWithFrame:(CGRectZero)];
        _inputTextView.font = [UIFont systemFontOfSize:16];
        _inputTextView.autoresizingMask = UIViewAutoresizingFlexibleHeight;
        _inputTextView.scrollEnabled = YES;
        _inputTextView.returnKeyType = UIReturnKeySend;
        _inputTextView.enablesReturnKeyAutomatically = YES; // UITextView内部判断send按钮是否可以用
        _inputTextView.backgroundColor = [UIColor clearColor];
        _inputTextView.keyboardType = UIKeyboardTypeDefault;
        [_inputTextView.undoManager disableUndoRegistration];
    }
    return _inputTextView;
}

- (UIButton *)imagePickButton {
    if (!_imagePickButton) {
        self.imagePickButton = [UIButton buttonWithType:(UIButtonTypeCustom)];
//        [_imagePickButton addTarget:self action:@selector(onImagePickButtonClick:) forControlEvents:(UIControlEventTouchUpInside)];
    }
    return _imagePickButton;
}

- (UIButton *)modeSwitchButton {
    if (!_modeSwitchButton) {
        _modeSwitchButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [_modeSwitchButton addTarget:self action:@selector(modeSwitchClick:) forControlEvents:UIControlEventTouchUpInside];
    }
    return _modeSwitchButton;
}

- (UIButton *)faceButton {
    if (!_faceButton) {
        _faceButton = [[UIButton alloc] initWithFrame:CGRectZero];
        _faceButton.selected = NO;
        _faceButton.autoresizingMask = UIViewAutoresizingFlexibleTopMargin | UIViewAutoresizingFlexibleLeftMargin;
//        [_faceButton addTarget:self action:@selector(didClickFaceBtn:) forControlEvents:UIControlEventTouchUpInside];
    }
    return _faceButton;
}

- (LRSMessageToolBarRecordButton *)recordingBtn {
    if (!_recordingBtn) {
        _recordingBtn = [LRSMessageToolBarRecordButton buttonWithType:UIButtonTypeCustom];
        _recordingBtn.backgroundColor = [UIColor clearColor];
        _recordingBtn.layer.masksToBounds = YES;

        _recordingBtn.clickTime = 0;
        _recordingBtn.hidden = YES;
        _recordingBtn.selected = NO;
    }
    return _recordingBtn;
}


@end

File: LRSMessageToolBar/Classes/InputBar/LRSMessageInputBar.swift
//
//  LRSMessageInputBar.swift
//  LRSMessageToolBarComponents
//
//  Created by sama 刘 on 2021/11/16.
//

import UIKit
import SnapKit

@objc public class LRSMessageInputBar: UIView {

    enum Mode {
        case input
        case record
    }

    private let configure: LRSMessageToolBarConfigure
    var mode: Mode = .input {
        willSet {
            switchMode(m: newValue)
        }
    }

    lazy var modeSwitchButton: UIButton = UIButton(type: .custom).then {
        $0.addTarget(self, action: #selector(modeSwitchClick(button: )), for: .touchUpInside)
    }

    @objc lazy var recordingBtn: LRSMessageToolBarRecordButton =  LRSMessageToolBarRecordButton(frame: .zero).then {
        $0.build(state: configure.recordButton.state)
        $0.isHidden = true
        $0.clickTime = configure.recordButton.clickTime
        $0.areaY = configure.textView.minHeight
    }

    @objc public lazy var inputTextView: LRSPlaceholderTextView = LRSPlaceholderTextView().then {
        $0.font = .systemFont(ofSize: 16)
        $0.autoresizingMask = .flexibleHeight
        $0.returnKeyType = .send
        $0.enablesReturnKeyAutomatically = true
        $0.backgroundColor = .clear
        $0.keyboardType = .default
        $0.undoManager?.disableUndoRegistration()
        $0.placeHolder = configure.textView.placeholder
        $0.placeHolderColor = configure.textView.placeHolderColor
    }

    lazy var faceButton: UIButton = UIButton(type: .custom).then {
        $0.build(state: configure.buttons.emojiButtonState)
    }

    @objc public lazy var imagePickButton: UIButton = UIButton(type: .custom).then {
        $0.build(state: configure.buttons.imagePickerButtonState)
    }

    lazy var bottomLine: UIView = UIView().then {
        $0.backgroundColor = .color(named: "Color_226")
    }

    public init(frame: CGRect, configure: LRSMessageToolBarConfigure = .default) {
        self.configure = configure
        super.init(frame: frame)
        loadBasicSubviews()
        switchMode(m: .input)
    }

    required public init?(coder: NSCoder) {
        fatalError()
    }

    private func loadBasicSubviews() {
        addSubview(imagePickButton)
        addSubview(faceButton)
        addSubview(inputTextView)
        addSubview(recordingBtn)
        addSubview(modeSwitchButton)
        addSubview(bottomLine)

        bottomLine.snp.makeConstraints { make in
            make.bottom.equalTo(inputTextView.snp.bottom)
            make.left.right.equalTo(inputTextView)
            make.height.equalTo(0.5)
        }
        modeSwitchButton.snp.makeConstraints { make in
            make.left.equalTo(configure.modeSwitch.leftMargin)
            make.size.equalTo(configure.modeSwitch.buttonSize)
            make.centerY.equalTo(inputTextView)
        }

        recordingBtn.snp.makeConstraints { make in
            make.left.equalTo(modeSwitchButton.snp.right).offset(configure.textView.leftMargin)
            make.top.equalTo(configure.textView.topMargin)
            make.right.equalTo(imagePickButton.snp.left).offset(-configure.textView.topMargin)
            make.height.equalTo(configure.textView.minHeight)
        }

        inputTextView.snp.makeConstraints { make in
            make.left.top.equalTo(recordingBtn)
            make.right.equalTo(faceButton.snp.left).offset(-configure.textView.topMargin)
            make.bottom.equalToSuperview()
        }

        imagePickButton.snp.makeConstraints { make in
            make.right.equalTo(-configure.buttons.rightButtonsMargin)
            make.centerY.equalTo(modeSwitchButton)
            make.width.height.equalTo(modeSwitchButton)
        };
        faceButton.snp.makeConstraints { make in
            make.size.equalTo(imagePickButton)
            make.centerY.equalTo(imagePickButton)
            make.right.equalTo(imagePickButton.snp.left).offset(configure.buttons.rightButtonsXSpacing)
        };
    }

    @objc private func modeSwitchClick(button: UIButton) {
        mode = mode == .input ? .record : .input
        if mode == .record {
            inputTextView.resignFirstResponder()
            recordingBtn.isSelected = false
        } else {
            inputTextView.becomeFirstResponder()
        }
    }

    private func switchMode(m: Mode) {
        recordingBtn.isHidden = m == .input
        modeSwitchButton.isSelected = !recordingBtn.isHidden
        inputTextView.isHidden = !recordingBtn.isHidden
        faceButton.isHidden = inputTextView.isHidden
        bottomLine.isHidden = LRSMessageToolBarHelper.safeAreaHeight == 0 || inputTextView.isHidden
        modeSwitchButton.setImage(configure.modeSwitch.state.image(for: m == .input ? .normal: .selected), for: .normal)
    }

}


fileprivate extension UIButton {
    func build(state: LRSMessageToolBarConfigure.ButtonState) {
        state.imageSource.forEach { key, value in
            setImage(value, for: key)
        }

        state.titleSource.forEach { key, value in
            setAttributedTitle(value, for: key)
        }
    }
}

fileprivate extension LRSMessageToolBarRecordButton {
    func build(state: LRSMessageToolBarConfigure.ButtonState) {
        state.imageSource.forEach { key, value in
            setImage(value, for: key)
        }

        state.titleSource.forEach { key, value in
            setTitle(value, for: key)
        }
    }
}
